services/src/tx_sender.rs

```diff
@@ -133,54 +78,36 @@ impl TxSender {
         notifier: Option<NotificationSender>,
     ) -> AnyhowJoinHandle {
         tokio::spawn(async move {
-            loop {
-                let mut transaction_infos = Vec::with_capacity(MAX_BATCH_SIZE_IN_PER_INTERVAL);
-                let mut timeout_interval = INTERVAL_PER_BATCH_IN_MS;
-
-                // In solana there in sig verify stage rate is limited to 2000 txs in 50ms
-                // taking this as reference
-                while transaction_infos.len() <= MAX_BATCH_SIZE_IN_PER_INTERVAL {
-                    let instance = tokio::time::Instant::now();
-                    match tokio::time::timeout(Duration::from_millis(timeout_interval), recv.recv())
-                        .await
-                    {
-                        Ok(value) => match value {
-                            Some(transaction_info) => {
-                                TXS_IN_CHANNEL.dec();
-
-                                // duplicate transaction
-                                if self
-                                    .data_cache
-                                    .txs
-                                    .contains_key(&transaction_info.signature)
-                                {
-                                    continue;
-                                }
-                                transaction_infos.push(transaction_info);
-                                // update the timeout inteval
-                                timeout_interval = timeout_interval
-                                    .saturating_sub(instance.elapsed().as_millis() as u64)
-                                    .max(1);
-                            }
-                            None => {
-                                log::error!("Channel Disconnected");
-                                bail!("Channel Disconnected");
-                            }
-                        },
-                        Err(_) => {
-                            break;
-                        }
+            let mut instant = Instant::now();
+            let mut notifications = vec![];
+            while let Some(transaction_info) = recv.recv().await {
```

Comment:
we can lock the process for an unknown time, better apply timeout + try_recv
something like this
```rust
loop {
    let deadline = Instant::now() + INTERVAL_PER_BATCH_IN_MS;
    let mut notifications = vec![];
    loop {
        let maybe_transaction_info = if deadline.elapsed().is_zer() {
            match tokio::time::timeout_at(deadline, recv.recv()).await {
                Ok(Some(tx_info)) => Some(tx_info),
                Ok(None) => bail!("Tx sender service stopped"),
                Err(_error) => None,
            }
        } else {
            match recv.try_recv() {
                tokio::sync::mpsc::error::TryRecvError::Empty => None,
                tokio::sync::mpsc::error::TryRecvError::Disconnected => bail!("Tx sender service stopped")
            }
        };

        if let Some(transaction_info) = maybe_transaction_info {
            self.forward_txs(&transaction_info).await;

            // send transaction notifications in batches
            if let Some(notifier) = &notifier {
                let forwarded_slot = self.data_cache.slot_cache.get_current_slot();
                let forwarded_local_time = Utc::now();
                let tx_notification = TransactionNotification {
                    signature: transaction_info.signature,
                    recent_slot: transaction_info.slot,
                    forwarded_slot,
                    forwarded_local_time,
                    processed_slot: None,
                    cu_consumed: None,
                    cu_requested: None,
                    quic_response: 0,
                };
                notifications.push(tx_notification);
            }
        } else {
            // send transaction notifications in batches
            if let Some(notifier) = &notifier {
                // send notification for sent transactions
                let _ = notifier.send(NotificationMsg::TxNotificationMsg(notifications));
            }
            break;
        }
    }
}
```

Reply:
Have addressed your concerns using tokio select, it seems more readable for me.